### PR TITLE
feat: sealing: storage redeclare/detach

### LINKED ISSUE #3716: Storage detach
Can we get a `lotus-miner storage **detach**` command?

### LINKED ISSUE #9019: lotus-miner/worker storage redeclare
Right now it's only possible to attach storage paths at runtime with `lotus-worker storage attach`.

If path config is changed, or sector files are moved around, there's no way to tell the sector index that something about paths has changed, so users have to restart the main instance of lotus-miner.

We should add commands which will let users refresh storage path metadata on lotus-miner.

This will be useful for things like applying path filters at runtime (https://github.com/filecoin-project/lotus/pull/9013)

Somewhat related to https://github.com/filecoin-project/lotus/issues/3716, might do them both at the same time

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues
Based on #9013
Closes #9019
Closes #3716

## Proposed Changes
1. Add new APIs on lotus-worker/lotus-miner
```go
StorageDetachLocal(ctx context.Context, path string) error                           //perm:admin
StorageRedeclareLocal(ctx context.Context, id *storiface.ID, dropMissing bool) error //perm:admin
```

2. Add an API on lotus-miner to tell the sector index that a path is being detached
```go
StorageDetach(ctx context.Context, id storiface.ID, url string) error                                                              //perm:admin
```

3. Add `storage detach/redeclare` commands to lotus-miner/lotus-worker
* `lotus-miner storage detach`
```
NAME:
   lotus-miner storage detach - detach local storage path
USAGE:
   lotus-miner storage detach [command options] [path]
OPTIONS:
   --really-do-it  (default: false)
   
```

* `lotus-miner storage redeclare`
```
NAME:
   lotus-miner storage redeclare - redeclare sectors in a local storage path
USAGE:
   lotus-miner storage redeclare [command options] [arguments...]
OPTIONS:
   --all           redeclare all storage paths (default: false)
   --drop-missing  Drop index entries with missing files (default: false)
   --id value      storage path ID
   
```

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

**Comment by magik6k** [comment]
Tests need fixing, likely due to changes in [itests ensemble: import presealed sector metadata](https://github.com/filecoin-project/lotus/pull/9032/commits/a1ece56058e46e1f3a06caa1acdcacca4dd12473) - a bunch of itests are likely calling `SectorsList`, and asserting things on those sectors, not expecting genesis sectors in that list.

Easiest fix for those is likely adding a helper method `SectorsListNonGenesis` on kit.TestMiner, and replacing uses of SectorsList in broken itests

File: cmd/lotus-miner/storage.go [hunk 1]
```diff
@@ -174,6 +176,86 @@ over time
...
+		if !cctx.Args().Present() {
+			return xerrors.Errorf("must specify storage path")
+		}
+
+		p, err := homedir.Expand(cctx.Args().First())
+		if err != nil {
+			return xerrors.Errorf("expanding path: %w", err)
+		}
+
+		if !cctx.Bool("really-do-id") {
```
Comment (rjan90) [inline cmd/lotus-miner/storage.go:205]: This seems like a typo #9086

**Comment by rjan90** [comment]
Been testing this PR, and after fixing the typo in `if !cctx.Bool("really-do-id")`, things are looking good so far:
- [x]  On a miner
   - [x] Detach a storage path (One small UX-nit here would be to print a "Detached /path/to/storage/ successfully" when detaching.
   - [x] Put sector files, and check that after redeclare, lotus-miner sectors find lists the sector in the path:

**Before move:**
```
lotus-miner storage find 2
In c3987105-1450-44fd-af25-ef15690c67a7 (Sealed, Cache)
        Sealing: false; Storage: true
        Local (/root/storage)
        URL: http://127.0.0.1:2345/remote/sealed/s-t01008-2
```
**Moving sectors:**
```
mv storage/sealed/s-t01008-2 storage45/sealed
mv storage/cache/s-t01008-2 storage45/cache
```
**Redeclare & check sectors find:**

```
lotus-miner storage redeclare --all /root/storage45
lotus-miner storage find 2
In c3987105-1450-44fd-af25-ef15690c67a7 (Sealed, Cache)
        Sealing: false; Storage: true
        Local (/root/storage)
        URL: http://127.0.0.1:2345/remote/sealed/s-t01008-2
In fb10368f-f33f-4882-a047-3e7f79412eaa (Sealed, Cache)
        Sealing: false; Storage: true
        Local (/root/storage45)
```

   - [x] Remove sector file(s), make sure they aren't listed in lotus-miner sectors find after lotus-miner storage redeclare --drop-missing

```
lotus-miner storage redeclare --drop-missing --all /root/storage
lotus-miner storage find 2
In fb10368f-f33f-4882-a047-3e7f79412eaa (Sealed, Cache)
        Sealing: false; Storage: true
        Local (/root/storage45)
        URL: http://127.0.0.1:2345/remote/sealed/s-t01008-2
```

Will go through the tests on a worker tomorrow.
- [ ] On a worker
   - [ ] Detach a storage path
   - [ ] Put sector files, and check that after redeclare, lotus-miner sectors find lists the sector in the path
   - [ ] Remove sector file(s), make sure they aren't listed in lotus-miner sectors find after lotus-worker storage redeclare --drop-missing

File: cmd/lotus-worker/storage.go [hunk 2]
```diff
@@ -128,3 +130,83 @@ var storageAttachCmd = &cli.Command{
...
+		if !cctx.Args().Present() {
+			return xerrors.Errorf("must specify storage path")
+		}
+
+		p, err := homedir.Expand(cctx.Args().First())
+		if err != nil {
+			return xerrors.Errorf("expanding path: %w", err)
+		}
+
+		if !cctx.Bool("really-do-id") {
```
Comment (rjan90) [inline cmd/lotus-worker/storage.go:160]: Similar typo here as in the `lotus-worker/storage` as the on in `lotus-miner/storage`

**Comment by rjan90** [comment]
Okay, so one strange thing happened during my testing of the `lotus-worker storage detach/attach` command.

I first detached the `/root/worker` storage path to test that it functioned, and after that I had to attach it again to test the redeclare function. After I attached the `/root/worker` path, that path is shown properly in the `lotus-worker info`:
```
lotus-worker info
Worker version:  1.6.0
CLI version: lotus-worker version 1.17.1-dev+butterflynet+git.459c5f008.dirty

Session: 9d5b00c5-f330-4447-b87a-c4556470519a
Enabled: true
Hostname: Ubuntu-2004-focal-64-minimal
CPUs: 12; GPUs: []
RAM: 4.781 GiB/62.79 GiB; Swap: 71.25 MiB/128 GiB
Task types: FIN GET FRU UNS C1 C2 PC2 PC1 PR1 PR2 RU AP DC GSK 

eb7d2640-20e6-4b2f-81ef-c60ea50c32d0:
        Weight: 10; Use: Seal 
        Local: /root/.lotusworker
2462c12e-ba04-489d-bd3c-b0df21cfb890:
        Weight: 10; Use: Seal 
        Local: /root/worker
```

But when I check the `lotus-miner storage list` that path is shown as not found 🤔. It was able to successfully move the finished sealed sector from the `/root/worker` path to the long-term storage path though.
```
007bc396-cf31-4493-b75a-412e8fdccc78:
        [##########################################        ] 366.8 GiB/436.3 GiB 84%
        Unsealed: 0; Sealed: 0; Caches: 0; Reserved: 0 B
        Weight: 10; Use: Seal 
        Local: /root/sealing
        URL: http://127.0.0.1:2345/remote

[...............]

2462c12e-ba04-489d-bd3c-b0df21cfb890:
        Error: fsstat: path not found:
```

**Comment by magik6k** [comment]
@rjan90 Managed to repro, and hopefully fix the issue in https://github.com/filecoin-project/lotus/pull/9032/commits/6d2990314f6148c8fbe0644e366144de5cf9011f

File: cmd/lotus-miner/storage.go [hunk 3]
```diff
@@ -174,6 +176,86 @@ over time
...
+		}
+
+		return nodeApi.StorageDetachLocal(ctx, p)
+	},
+}
+
+var storageRedeclareCmd = &cli.Command{
+	Name:  "redeclare",
+	Usage: "redeclare sectors in a local storage path",
+	Flags: []cli.Flag{
```
Comment (geoff-vball) [inline cmd/lotus-miner/storage.go:216]: add `ArgsUsage`

File: cmd/lotus-worker/storage.go [hunk 4]
```diff
@@ -128,3 +130,83 @@ var storageAttachCmd = &cli.Command{
...
+		},
+		&cli.BoolFlag{
+			Name:  "all",
+			Usage: "redeclare all storage paths",
+		},
+		&cli.BoolFlag{
+			Name:  "drop-missing",
+			Usage: "Drop index entries with missing files",
+		},
+	},
```
Comment (geoff-vball) [inline cmd/lotus-worker/storage.go:184]: ArgsUsage here too

**Review by geoff-vball — COMMENTED** [review]
Two quickly fixed comments.

If @rjan90 can confirm the fix you made, this should be good to go.

**Comment by magik6k** [comment]
Dropped the args check in the redeclare command as it doesn't use those args.

**Comment by rjan90** [comment]
> If @rjan90 can confirm the fix you made, this should be good to go.

I re-ran the detach shared paths issue, and where able to successfully detach the storage on both worker and miner now 👍 So I can confirm that it´s fixed.

**Review by geoff-vball — APPROVED** [review]

**Comment by jennijuju** [comment]
@magik6k conflicts on docs - then merge? we want this in v1.17.1!!

**Comment by lbj2004032** [comment]
This issue has not been resolved yet

Many nodes are stuck in the "New sector" log line

When I modify storage.json and change the Path to a new value (/mnt/lotus/mainData/), this Path is executed by re executing the lotus miner storage attach -- init -- store/mnt/lotus/mainData/, which can be started successfully

Next, I execute the lotus miner attach/mnt/netdisk/mainData command. The program is stuck, and the path/mnt/netdisk/mainData is the old value.

"I have to start using 1.17.0 first. If it succeeds, it must be started using 1.17.0. If it is 1.17.2, the program still cannot be started."

**Pull request merged**